mlx/ops.cpp

```diff
@@ -718,6 +717,39 @@ array stack(const std::vector<array>& arrays, StreamOrDevice s /* = {} */) {
   return stack(arrays, 0, s);
 }
 
+/** array repeat with axis */
+array repeat(const array& arr, int repeats, int axis, StreamOrDevice s) {
+  axis = normalize_axis(axis, arr.ndim());
+
+  if (repeats <= 0) {
+    std::vector<int> new_shape(arr.shape());
+    new_shape[axis] = repeats > 0 ? repeats : 0;
+    return zeros(new_shape, arr.dtype(), s);
+  }
+
+  if (repeats == 1) {
+    return arr;
+  }
+
+  std::vector<array> arrays_to_concat;
+  arrays_to_concat.reserve(repeats * arr.shape(axis));
+
+  for (int i = 0; i < arr.shape(axis); ++i) {
+    std::vector<int> start_indices(arr.ndim(), 0);
+    std::vector<int> stop_indices = arr.shape();
+    start_indices[axis] = i;
+    stop_indices[axis] = i + 1;
+    for (int j = 0; j < repeats; ++j) {
+      arrays_to_concat.push_back(slice(arr, start_indices, stop_indices, s));
+    }
+  }
```

Comment:
This needs to be implemented with a reshape and a concatenate that is order `repeats` and critically does not scale with the shape of the repeating axis. For example if you are repeating:

```
a = array([[0, 1], [2, 3]])
```

along axis `0`  (in numpy like psuedo-code) do:
```
concatenate([a]*repeats, -1).reshape(-1, a.shape[1])
```

and along axis `1` do:

```
stack([a]*repeats, -1).reshape(a.shape[0], -1)
```

Reply:
hey just wrap my head around this, do you mean I should handle the cases of 0 and 1 axis repetitions alone and 
the other axis with the same logic or for the other axis with only a reshape and a concatenate that is order repeats

Reply:
Yea, what I'm saying is we should avoid ever making `repeats * array.shape(axis)` subarrays to concatenate using something like the strategy I outlined above for a 2D array which just makes `repeats` arrays. I think it should generalize to ND arrays but let me know if I'm missing something.

Reply:
Yes we definitely can generalize this for ND array using something like this for axis 0 or 1 
```
def repeat(arr, repeats, axis):
    new_shape = np.array(arr.shape)
    new_shape[axis] *= repeats
    if axis == 0:
        repeated = np.concatenate([arr]*repeats, axis=-(len(arr.shape)-1)).flatten()
        return repeated.reshape(new_shape)
    elif axis == 1:
        repeated = np.stack([arr for _ in range(repeats)], axis=-(len(arr.shape)-1)).flatten()
        return repeated.reshape(new_shape)
```
 but what I'm trying to understand is if you think there is a better way for other than making slices for axis bigger than 0 or 1 for bigger arrays.

Reply:
Yea exactly, I think it should work for any axis of any n-darray. You have to concatenate along the correct dimension followed by a reshape.

Reply:
Hey please check the new commit it should be exactly what you suggested

Reply:
Much better thanks!! I left a few more comments. I think we should be good to go after you address them.